reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,807 @@
+ray.init()
+
+# The job can use at least 1 CPU and at most 8 CPUs.
+ray.init(virtual_cluster=VirtualCluster(
+    flex_resource_min={"CPU": 1}, flex_resource_max={"CPU": 8}))
+
+# The job uses 1 A100 GPU.
+ray.init(virtual_cluster=VirtualCluster(
+    fixed_size_nodes=[
+        FixedSizeNode(
+            resources={"GPU": 1},
+            parent_node_selector={"accelerator_type": In("A100")}
+        )
+    ]))
+```
+
+```shell
+# Default virtual cluster
+# The job can use up to the entire physical cluster resources.
+ray job submit -- python job.py
+
+# The job needs 2 * 1 GPU that are strict spreaded.
+ray job submit --virtual-cluster='{"fixed_size_nodes": [{"nodes": [{"resources": {"GPU": 1}}, {"resources": {"GPU": 1}}], "scheduling_policy": "STRICT_SPREAD"}]}' -- python job.py
+```
+
+Once a job is running inside a virtual cluster, it can use all the Ray APIs as if it's running inside its own Ray cluster.
+
+#### Placement Group API
+
+Since virtual clusters are nestable and support gang scheduling, they can be used to implement or replace placement groups.
+
+```python
+# Create a placement group with two bundles that are packed
+pg = placement_group([{"GPU": 4}, {"GPU": 1}], strategy="PACK")
+# Run the task inside the placement group
+task.options(
+    scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=pg)).remote()
+
+# Create a virtual cluster with two virtual nodes that are packed
+vc = VirtualCluster(
+    fixed_size_nodes=[
+        FixedSizeNodes(
+            nodes=[
+                FixedSizeNode(resources={"GPU": 4}),
+                FixedSizeNode(resources={"GPU": 1})
+            ],
+            scheduling_policy=PACK
+        )
+    ])
+# A nested virtual cluster inside the job virtual cluster
+with vc:
+  # Run the task inside the virtual cluster
+  task.remote()
+```
+
+```python
+# Create a placement group with two bundles that are strict spreaded
+pg = placement_group([{"GPU": 4}, {"GPU": 1}], strategy="STRICT_SPREAD")
+# Run the actor using the bundle 0 resources
+Actor.options(
+    scheduling_strategy=PlacementGroupSchedulingStrategy(
+        placement_group=pg, placement_group_bundle_index=0)).remote()
+
+# Create a virtual cluster with two virtual nodes that are strict spreaded
+vc = VirtualCluster(
+    fixed_size_nodes=[
+        FixedSizeNodes(
+            nodes=[
+                FixedSizeNode(resources={"GPU": 4}, labels={"bundle_index": "0"}),
+                FixedSizeNode(resources={"GPU": 1}, labels={"bundle_index": "1"})
+            ],
+            scheduling_policy=STRICT_SPREAD
+        )
+    ])
+with vc:
+  # Run the actor using the first virtual node resources
+  Actor.options(node_labels={"bundle_index": In("0")}).remote()
+```
+
+Here we have two options for the API:
+
+1. Deprecate the placement group API and use virtual cluster API directly.
+2. Keep the placement group API and only change the internal implementation to use virtual cluster.
+
+#### Examples
```

Comment:
Another example to consider:

#### ML platform with priority-based pre-emption of distributed training workers

```
Overall Ray Cluster: 16 GPUs, 128 CPUs
    Job 0 (P0): min_gpus=4, max_gpus=10, cpus=32
        Worker * 10: each with 1 GPU
    Job 1 (P1): min_gpus=2, max_gpus=8, cpus=1
        Worker * 3: each with 1 GPU
    Job 2 (P1): min_gpus=1, max_gpus=4, cpus=1
        Worker * 3: each with 1 GPU
    Job 3 (P2): min_gpus=1
        Pending for something to free up. 
```

It might make more sense to switch these priorities to GPU time in %. This arbitrary upscale/downscale depends on elastic distributed training from Ray Train / another library.

This might be a little too much scope for the current REP, but I would like a future REP to enable this kind of pre-emption policy for each "parent" virtual cluster. The parent VirtualCluster should be able to drain some nodes of the child clusters to reclaim resources, as a form of "hard downscaling", rather than just waiting for task completion.